refactor: create problem feedback runtime service

Cup0fCoffee opened the pull request:
<!--

🌳🌳
🌳🌳🌳🌳         🌳 Note: Quince is in support. Fixes you make on master may still be needed on Quince.
    🌳🌳🌳🌳     If so, make another pull request against the open-release/quince.master branch,
🌳🌳🌳🌳         or ask in the #wg-build-test-release Slack channel if you have any questions or need help.
🌳🌳


Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly
readable.  If you must linked information must be private (because it has secrets),
clearly label the link as private.

-->

## Description

This PR resolves https://github.com/openedx/edx-platform/issues/33325.

The original PR - https://github.com/openedx/edx-platform/pull/33424 - contains most of the context and discussion. We had to duplicate the PR, as the original one was open from a fork that OpenCraft doesn't have necessary permissions to be able to squash and rebase commits.

## Supporting information

Link to other information about the change, such as Jira issues, GitHub issues, or Discourse discussions.
Be sure to check they are publicly readable, or if not, repeat the information here.

## Testing instructions

* Checkout to the current branch and build 
* Browse the demo course to Content > Outline and search the Homework task
* Pick any of the activities in the live and studio version and use the show answer button

## Deadline

"None"

steff456 committed:
refactor: create problem feedback runtime service

openedx-webhooks commented:
Thanks for the pull request, @Cup0fCoffee!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

Cup0fCoffee committed:
fix: failing tests and edge cases related to due date

refactor: check if xblock is None

fix: add minor fixes and helpful comments

fix: circular imports

Cup0fCoffee commented:
@Agrendalath When I saw the tests failing, I realized that there some bugs and tried fixing them first. But then I decided to check the original PR, and found that Stephannie [brought this up already](https://github.com/openedx/edx-platform/pull/33424#discussion_r1398104145), and David [has replied](https://github.com/openedx/edx-platform/pull/33424#discussion_r1401441950) that since this is a refactoring, the tests should not be changed, even though it does look like a bug. So I made changes to the implementation in the latest commit (didn't squash yet, so it's easy to see what changes were made) to make sure that the tests are passing. The tests that I did modify, are the ones where we actually changed the implementation of the functionality that it's touching.

Feel free to squash and merge, if you're ok with the changes.

Agrendalath commented on xmodule/services.py:
```diff
@@ -308,3 +312,133 @@ def _handle_deprecated_progress_event(self, block, event):
+        """
+        if self._xblock:
+            due_date = getattr(self._xblock, 'close_date', None) or getattr(self._xblock, 'due', None)
+            return (due_date is None or datetime.now(UTC) > due_date)
```
Context: we changed this condition from the following:
```python
return (self.close_date is not None and
        datetime.datetime.now(utc) > self.close_date)
```

Why do we want to return `True` when there is no `close_date` or `due` set? Won't it affect the following code?
https://github.com/openedx/edx-platform/blob/b6634a811abf03f2c3b39fff10b0c56c50c9e78c/openedx/features/personalized_learner_schedules/call_to_action.py#L80-L84

Besides, is there a reason to avoid using the methods and properties already provided by the `InheritanceMixin`?
https://github.com/openedx/edx-platform/blob/f544a4825dffeef684401f4a7cf0f68b9c4b3549/xmodule/modulestore/inheritance.py#L250-L269

Cup0fCoffee replied:
This is related to the [comment I left above](https://github.com/openedx/edx-platform/pull/34320#issuecomment-1976850180).

One on hand, we want it to be `false` when there is no due date set (I'll ignore `close_date`, because it's `due + graceperiod`, so if `due` is `None`, `close_date` is also `None`), for example when we want the learner to be able to complete problems in the self paced course where there are no due dates. On another hand, there is logic related to displaying the answer and it's correctness, that depends on the due date, and in the scenario when there is no due date, we want the learner to be able to see the answer right after they've completed the problem, and in that case we want to it to return `true`. So these to interpretations of the `is_past_due` are conflicting. We could keep this method internal to this service.

Re `InheritanceMixin`, I tried removing the `is_past_due` method of the capa XBlock, and the tests started failing, so I decided to play it save, and keep it there, as I was trying to make the least number of changes to an already approved PR.

Cup0fCoffee replied:
Also, the reason I decided to check for both `close_date` and `due` properties of the XBlock, is because on the code block referenced below, XBlocks that are passed to the service seem to not have an implementation for `close_date`, because the tests start failing. That's why I added the fallback to the `due` property.

https://github.com/openedx/edx-platform/blob/5907b7e4f5ffa18dba4598871816b9fa353f8dcf/lms/djangoapps/grades/subsection_grade.py#L60-L64

~I'm not sure why the don't inherit from the `InheritanceMixin`, but if they were, they would have the `close_date` property.~ It seems like the `BlockFactory` creates XBlocks that don't inherit from this mixin.

You can see failed test [here](https://github.com/openedx/edx-platform/actions/runs/8141411936/job/22248710573).

Cup0fCoffee replied:
I tried removing the `is_past_due` method of the `ProblemXBlock`, located the tests that are failing (bellow for reference), and found that the in the factory that creates the xblocks for the tests the line that should add these mixins is commented out. Do you know why it was commented out? And whether we can uncomment them

```
================================================================================================================================= short test summary info ==================================================================================================================================
FAILED xmodule/tests/test_capa_block.py::ProblemBlockTest::test_submit_problem_closed - AttributeError: 'ProblemBlock' object has no attribute 'is_past_due'
FAILED xmodule/tests/test_delay_between_attempts.py::XModuleQuizAttemptsDelayTest::test_still_cannot_submit_after_max_attempts - AttributeError: 'ProblemBlock' object has no attribute 'is_past_due'
```

https://github.com/openedx/edx-platform/blob/5907b7e4f5ffa18dba4598871816b9fa353f8dcf/xmodule/tests/__init__.py#L239-L243

Though uncommenting these lines doesn't fix the issue.

Agrendalath replied:
@Cup0fCoffee, I pushed 184516500f3a56860b4b26c98ecfc379e91eb025, which initializes XBlock mixins in CAPA tests.

Short explanation:
1. `CapaFactory` creates a new ProblemBlock. As the first step, it initializes the runtime here:
   https://github.com/openedx/edx-platform/blob/184516500f3a56860b4b26c98ecfc379e91eb025/xmodule/tests/test_capa_block.py#L121-L125
   (Note: I renamed "system" to "runtime" because the "system" is [a deprecated term](https://github.com/openedx/edx-platform/blob/184516500f3a56860b4b26c98ecfc379e91eb025/openedx/core/djangoapps/xblock/runtime/shims.py#L321-L331).)
2. The `get_test_system` function runs `get_test_descriptor_system` to create a new `TestDescriptorSystem`. As we can see below, this class already contains the `InheritanceMixin` during initialization:
   https://github.com/openedx/edx-platform/blob/184516500f3a56860b4b26c98ecfc379e91eb025/xmodule/tests/__init__.py#L259
3. The `mixins` arg passed there is then used during the initialization of the `Runtime` class. It creates [a new `Mixologist` instance](https://github.com/openedx/XBlock/blob/bd289d2f9f17c62276df24d8a57b437b160114ec/xblock/runtime.py#L568).
4. However, the mixing (`mixologist.mix(cls)(...)`) needs to be explicitly executed. As it's implicitly executed in the [`construct_xblock_from_class` method](https://github.com/openedx/XBlock/blob/bd289d2f9f17c62276df24d8a57b437b160114ec/xblock/runtime.py#L641-L651), I decided to use it instead of initializing the block directly from the `ProblemBlock` class.

Cup0fCoffee replied:
@Agrendalath Nice! Thank you for the explanation. I found another place in xmodule tests that needs this change, and applied it.

I've also removed the `close_date` and `is_past_due` methods from the `ProblemXBlock`.

Agrendalath commented on xmodule/services.py:
```diff
@@ -308,3 +312,133 @@ def _handle_deprecated_progress_event(self, block, event):
+        if self._xblock:
+            show_correctness = self._xblock.show_correctness
+        else:
+            show_correctness = ''
```
Shouldn't we just return `False` at this point? The current implementation compares an empty string to the constants defined in the `ShowAnswer` class. Ultimately, it returns `True` despite not matching any of these values.

Cup0fCoffee replied:
Strictly speaking, isn't it the same as the original implementation? I.e. would always return `True` when `show_correctness` is `''`, and also we would be ignoring the `is_staff` check.

https://github.com/openedx/edx-platform/blob/794c6781fad8d290b548392a94f31a2d1b18e86f/xmodule/graders.py#L487-L503

I'm not against this suggestion. I was just trying to follow David's advice, i.e. to avoid changing the behavior, since it's a refactoring PR.

Agrendalath commented on xmodule/services.py:
```diff
@@ -308,3 +312,133 @@ def _handle_deprecated_progress_event(self, block, event):
+            max_attempts = 0
+            attempts = 0
+            required_attempts = 0
+            is_correct = False
```
Do we need this? It sets `show_answer = ''`, then it goes to `self.correctness_available()` (which returns `True` as described in another comment), and finally, it compares the empty string (`show_answer`) to each constant from the `ShowAnswer` and returns `False` because it did not match any of them. The result seems to be identical every time, so both this and `correctness_available` could have something like this at the beginning:
```python
if not self._xblock:
    return False
```

Cup0fCoffee replied:
You're right. I'll refactor all methods of this service to use this check for consistency.

Agrendalath commented on xmodule/tests/test_services.py:
```diff
@@ -163,3 +166,84 @@ class TestTeamsConfigurationService(ConfigurationServiceBaseClass):
+            self._xblock.close_date = None
+        else:
+            self._xblock.due = getattr(self, due_date_str)
+            self._xblock.close_date = getattr(self, due_date_str)
```
The `closed_date` property should be provided by the `InheritanceMixin`. Isn't that the case here?

Cup0fCoffee replied:
In this case `self._xblock` is an instance of `MagicMock` class.

https://github.com/openedx/edx-platform/blob/5907b7e4f5ffa18dba4598871816b9fa353f8dcf/xmodule/tests/test_services.py#L172-L184

Agrendalath committed:
test: use InheritanceMixin in ProblemBlock tests

Cup0fCoffee committed:
refactor: remove ProblemXBlock's methods that are already inherited

close_date and is_past_due methods are already implemeneted in
InheritanceMixin. All of the XBlock should already be inheriting from
this class when constructured by XBlock runtime, so there is no reason
for duplicating these methods.

Cup0fCoffee committed:
feat: add InheritanceMixin to list of mixins of XBlockRuntime

Necessary for both - test and production runtimes.

Agrendalath commented on xmodule/services.py:
```diff
@@ -17,13 +20,14 @@
 from lms.djangoapps.lms_xblock.models import XBlockAsidesConfig
 
-from lms.djangoapps.grades.api import signals as grades_signals
+from lms.djangoapps.grades.signals import signals as grades_signals
```
@Cup0fCoffee, why did we change this one? The idea is to import things from other apps through the `api` module.

Cup0fCoffee replied:
@Agrendalath I didn't dig too much into it (I'm going to dig a bit more after I post this comment), but I think it's because due to some `import` changes, a circular imports occur. I'm getting a following error when I change from `signals` to `api`:

```
Traceback (most recent call last):
  File "manage.py", line 103, in <module>
    startup.run()
  File "/edx/app/edxapp/edx-platform/lms/startup.py", line 20, in run
    django.setup()
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/apps/registry.py", line 116, in populate
    app_config.import_models()
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/apps/config.py", line 269, in import_models
    self.models_module = import_module(models_module_name)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/bulk_email/models.py", line 22, in <module>
    from openedx.core.djangoapps.course_groups.cohorts import get_cohort_by_name
  File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/course_groups/cohorts.py", line 21, in <module>
    from lms.djangoapps.courseware import courses
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/courses.py", line 39, in <module>
    from lms.djangoapps.courseware.date_summary import (
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/date_summary.py", line 21, in <module>
    from lms.djangoapps.certificates.api import get_active_web_certificate, can_show_certificate_available_date_field
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/certificates/api.py", line 27, in <module>
    from lms.djangoapps.certificates.generation_handler import (
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/certificates/generation_handler.py", line 24, in <module>
    from lms.djangoapps.grades.api import CourseGradeFactory
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/api.py", line 15, in <module>
    from lms.djangoapps.grades import constants, context, course_data, events
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/context.py", line 10, in <module>
    from .course_grade import CourseGrade
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/course_grade.py", line 17, in <module>
    from .subsection_grade import ZeroSubsectionGrade
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/subsection_grade.py", line 16, in <module>
    from xmodule.services import ProblemFeedbackService
  File "/edx/app/edxapp/edx-platform/xmodule/services.py", line 30, in <module>
    from lms.djangoapps.grades.api import signals as grades_signals
ImportError: cannot import name 'signals' from partially initialized module 'lms.djangoapps.grades.api' (most likely due to a circular import) (/edx/app/edxapp/edx-platform/lms/djangoapps/grades/api.py)
```

Cup0fCoffee replied:
From the stack trace above, it can be observed that when importing from `grades.api`, it imports other modules from `grades`. The `subsection_grade` module tries to import from `xmodule.services`, which in turn tries to import from `grades.api`, which forms the circular dependency.

Cup0fCoffee replied:
In the `subsection_grade` the `ProblemFeedbackService`, which is imported from the `xmodule.services` is used only once in one of the methods. We can move the import from the top of the file into the method body, to avoid the circular dependency. That way the `xmodule.services` will be able to import from the `grades.api`. I'll push the commit now, but please let me know if you think that's not a good approach.

Agrendalath commented on openedx/core/djangoapps/xblock/runtime/runtime.py:
```diff
@@ -101,6 +102,7 @@ def __init__(self, system: XBlockRuntimeSystem, user: UserType | None):
             mixins=(
                 LmsBlockMixin,  # Adds Non-deprecated LMS/Studio functionality
                 XBlockShim,  # Adds deprecated LMS/Studio functionality / backwards compatibility
+                InheritanceMixin,
```
@Cup0fCoffee, I need some help understanding the implications of this change.

@bradenmacdonald, @ormsbee, @kdmccormick, do you know why the `InheritanceMixin` was not explicitly listed in the mixins for the new runtime?

bradenmacdonald replied:
> do you know why the InheritanceMixin was not explicitly listed in the mixins for the new runtime?

The new runtime is only used for content libraries and LabXchange, both of which have very flat content trees, so I think that we didn't have any need for it.

mphilbrick211 replied:
Hey @bradenmacdonald @Cup0fCoffee @Agrendalath! Just checking in on this.

kdmccormick replied:
There is a long-term interest in flattening out XBlocks. All XBlocks would be leaf nodes, and structure/dynamicism would be handled by separate subsystem(s) in defined in openedx-learning. Leaving InheritanceMixin out of the new runtime was a nice nod towards this. This [ADR describes the rationale](https://github.com/openedx/edx-platform/blob/master/docs/decisions/0006-role-of-xblock.rst), although it only goes as far as saying that the Unit should be the highest-level xblock, rather than getting rid of inheritance entirely.

I know there are still several blocks today that rely on parent-child relationships: chapter/sequence (obviously), library_content, split_test, conditional, problem_builder. So, it might be inevitable that we add InhertianceMixin the new runtime. @Cup0fCoffee can you help me understand why this PR in particular needs to add the mixin?

Cup0fCoffee replied:
@kdmccormick The reasoning behind it is that @Agrendalath noticed that `ProblemBlock` implements several methods that are also implemented in the `InheritanceMixin` exactly the same way. Since there was an assumption that XBlocks inherit from the `InheritanceMixin`, the methods that already implemented by the `InheritanceMixin` can be removed from the `ProblemBlock`. So in this PR I attempted to remove the `close_date` and `is_past_due` methods.

Once I removed the methods, some tests started failing with errors like `AttributeError: 'ProblemBlock' object has no attribute 'is_past_due'`, because runtimes in those tests were not adding `InheritanceMixin` to the list of mixins, so the XBlocks that were being created were not inheriting from it. So for each failing test, I started tracking down the runtimes that were being used, and adding the mixin to the list until all the tests with this specific error started passing.

You can find more details in the following thread - https://github.com/openedx/edx-platform/pull/34320#discussion_r1519968408.

kdmccormick replied:
Thanks for you explanation @Cup0fCoffee , and sorry it took so long for me to get back to you.

I was confused for a while because I had forgotten that InheritanceMixin doesn't actually add inheritance capabilities... rather, it just _defines fields which the modulestore runtime treats as inheritable_.

Given that, I am OK with adding the mixin to Learning Core runtime, but with the knowledge that no field inheritance will actually happen in that runtime. Could you add a note to InheritanceMixin explaining this? Even better, if you have time, I think it would be good to rename the mixin to InheritableFieldsMixin.

That's my sense at least. @bradenmacdonald , does that sound right to you too?

bradenmacdonald replied:
That sounds reasonable @kdmccormick.

mphilbrick211 commented:
Hi @Cup0fCoffee @kdmccormick @bradenmacdonald is this still in progress?

mphilbrick211 commented:
> Hi @Cup0fCoffee @kdmccormick @bradenmacdonald is this still in progress?

Friendly ping on this @Cup0fCoffee @kdmccormick @bradenmacdonald

bradenmacdonald commented:
@Cup0fCoffee @Agrendalath Can you advise - is this something you want to get merged?

Cup0fCoffee commented:
@bradenmacdonald I guess, it would make sense to finish it, assuming we are ok spending more time on this (iirc, there isn't much left to do here). I'm not the assignee on the internal ticket, but I can assign it to myself, and put it in the next sprint.

Cup0fCoffee commented:
@bradenmacdonald Could you help me understand [this change](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/xblock/runtime/shims.py#L331-L340) you've done a while back?

In the context of this PR, I've noticed the tests are failing because of it. Via `git bisect`, I've identified that the commit on which the tests start failing is https://github.com/openedx/edx-platform/pull/34320/commits/121f9b973ce955baf4f1baabb1a5fdc761339ffb, i.e. the one that adds `InheritanceMixin` to the runtime. That makes sense, because now the XBlocks that inherit from that mixin have that field, so when trying to sync the fields, `graded` property is requested via `getattr` (I don't think the details are important so I'm not leaving any references). But because `XBlockShim` comes first in the inheritance list, it's property method is called first, which raises the error. This can be easily fixed by re-arranging the list and putting the `InheritanceMixin` after `XBlockShim`, but I think this would only hide the problem you were trying to highlight.

The comment on the method (btw, thank you for leaving it, it helps understanding the context) says:
```
Not sure what this is or how it's supposed to be set. Capa seems to expect a 'graded' attribute to be present on itself. Possibly through contentstore's update_section_grader_type() ?
```

I think there are two main points in that comment:
1. How the value is set?
2. It's only needed by Capa.

I think all XBlocks other than Capa were supposed to inherit this field from `InheritanceMixin`, and when I grep for `\.graded\b`, I see many references that don't seem exclusive to Capa.

Re how the value is set, my assumption is that it's done primarily through UI. In legacy Studio this field was probably automatically rendered. I did struggle to find relevant code in the authoring MFE that would set a value for this field. However, after that I found code for changing "grader type" ([here](https://github.com/openedx/frontend-app-authoring/blob/dc0ba6aac4bc1218885d45fe1fe12d6f956fbe15/src/generic/configure-modal/BasicTab.jsx#L66) and [here](https://github.com/openedx/frontend-app-authoring/blob/master/src/generic/configure-modal/ConfigureModal.jsx#L188)), and then figured out that `.graded` is set to `True` for all grader types except for `notgraded` [here](https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/models/settings/course_grading.py#L281).

In conclusion, I think that `.graded` property is needed, and not only for capa problems. Do you think the same, or is there something I've overlooked? If you do agree, can I remove the `graded` method that is raising the error from `XBlockShim`?

Cup0fCoffee committed:
refactor: move course_end_date property to InheritanceMixin

Cup0fCoffee committed:
refactor: rename InheritanceMixin to InheritableFieldsMixin

bradenmacdonald commented:
@Cup0fCoffee 

> In conclusion, I think that .graded property is needed, and not only for capa problems. Do you think the same, or is there something I've overlooked? If you do agree, can I remove the graded method that is raising the error from XBlockShim?

That makes sense. I think my confusion around `graded` is simply because I didn't realize it came from a mixin that we weren't using in the new runtime. I only added it to the shim because capa problems weren't working without it when I tested them, but everything else was working fine. It makes sense that with the InheritanceMixin, more things would require it.

So yes, please feel free to remove `graded` from the shim if you you are adding it via `InheritanceMixin`.

Cup0fCoffee committed:
refactor: remove graded property shim

graded property method is going to be inherited by XBlocks from the
InheritanceMixin added to runtime.

Cup0fCoffee commented:
@Agrendalath @bradenmacdonald @kdmccormick I've addressed all the feedback and all the tests are passing. This is ready for review, and hopefully, merging.

I've squashed and rearranged commits in the way I see it best, but if you think squashing them all into one commit is better, please let me know.

bradenmacdonald commented:
> Checkout to the current branch and build
> Browse the demo course to Content > Outline and search the Homework task
> Pick any of the activities in the live and studio version and use the show answer button

I don't see any section called "Homework" in the current version of the demo course. Maybe it's the one now called "Basic Assessment Tools" ?

bradenmacdonald commented:
In any case, this PR looks good to me at a very high level, and I know it's been well reviewed already. I read through the code and didn't see any concerns. (Though I would have preferred to have the InheritanceMixin rename split out into a separate PR, since that's easy enough to approve.)

Anyhow, @Agrendalath and/or @ormsbee do you want to do any further review before we merge this PR? Is it ready as far as you know?

Agrendalath commented:
@Cup0fCoffee, @bradenmacdonald, I don't remember much about this PR, but I can review it again in the next sprint.

Agrendalath commented on xmodule/modulestore/inheritance.py:
```diff
@@ -30,8 +30,12 @@ def to_json(self, values):  # lint-amnesty, pylint: disable=arguments-differ
 
-class InheritanceMixin(XBlockMixin):
-    """Field definitions for inheritable fields."""
+class InheritableFieldsMixin(XBlockMixin):
```
I know this change was requested [here](https://github.com/openedx/edx-platform/pull/34320#discussion_r1686874633), but this will break the import in [edx-when](https://github.com/openedx/edx-when/blob/83562f3/edx_when/field_data.py#L12). Even though `edx-when` provides a fallback for such a scenario, we should either keep the `InheritanceMixin` name for backward compatibility (with a deprecation warning) or add support for both names in `edx-when`.

Nit: there are also mentions of the `InheritanceMixin` in the docstrings of the [DnDv2 XBlock](https://github.com/openedx/xblock-drag-and-drop-v2/blob/f896375/drag_and_drop_v2/drag_and_drop_v2.py#L706) and [SGA XBlock](https://github.com/mitodl/edx-sga/blob/798022/edx_sga/sga.py#L933).

Agrendalath commented on xmodule/modulestore/inheritance.py:
```diff
@@ -30,8 +30,12 @@ def to_json(self, values):  # lint-amnesty, pylint: disable=arguments-differ
+    """
+    Field definitions for inheritable fields.
+
+    Defines fields which the modulestore runtime treats as inheritable.
```
What does it mean for the non-modulestore runtime?

Agrendalath commented on cms/djangoapps/contentstore/utils.py:
```diff
@@ -1279,7 +1284,8 @@ def load_services_for_studio(runtime, user):
         "teams_configuration": TeamsConfigurationService(),
-        "library_tools": LegacyLibraryToolsService(modulestore(), user.id)
+        "library_tools": LegacyLibraryToolsService(modulestore(), user.id),
+        "problem_feedback": ProblemFeedbackService,
```
Do we need it here? I thought we used this service only when rendering the XBlock or its preview, not in the Studio editor.

Agrendalath commented on xmodule/capa_block.py:
```diff
@@ -1438,25 +1400,13 @@ def hint_button(self, data):
-        """
-        return (self.close_date is not None and
-                datetime.datetime.now(utc) > self.close_date)
+        return self.runtime.service(self, 'problem_feedback').used_all_attempts()
```
Is this used anywhere? The only usages of the `used_all_attempts` method I found were directly in the `ProblemFeedbackService`.

Agrendalath commented on xmodule/services.py:
```diff
@@ -308,3 +312,147 @@ def _handle_deprecated_progress_event(self, block, event):
+        if not self._xblock:
+            return False
+
+        due_date = getattr(self._xblock, 'close_date', None) or getattr(self._xblock, 'due', None)
```
The `close_date` method already checks `self.due`. Why are we checking `block.due` separately here? Is there a use case where an XBlock does not inherit from the `InheritableFieldsMixin` but uses this service?

Agrendalath commented on xmodule/services.py:
```diff
@@ -308,3 +312,147 @@ def _handle_deprecated_progress_event(self, block, event):
+        when an XBlock doesn't have a due date, for example, when determining
+        whether a learner should still be able to submit their answer.
+        """
+        if not self._xblock:
+            return False
```
While raising an exception here would cause a runtime error affecting the learners, this check can result in a silent failure that's hard to detect. I wonder when this scenario (i.e., missing XBlock) can happen. Should we log an error in such cases?

Agrendalath commented on xmodule/services.py:
```diff
@@ -308,3 +312,147 @@ def _handle_deprecated_progress_event(self, block, event):
+        if not self._xblock:
+            return False
+
+        return self._xblock.attempts > 0
```
`attempts`  and `is_correct` are fields/properties specific to the Problem XBlock, unlike fields like `showanswer`, `due`, `show_correctness`, `max_attempts` that are provided by the `InheritableFieldsMixin`. We don't need to move this field to the mixin in this PR but we should at least add a note about it, because using this service with any other XBlock will result in an `AttributeError`.

Agrendalath commented on xmodule/services.py:
```diff
@@ -308,3 +312,147 @@ def _handle_deprecated_progress_event(self, block, event):
+        max_attempts = self._xblock.max_attempts
+        attempts = self._xblock.attempts
+        is_correct = self._xblock.is_correct()
+        required_attempts = self._xblock.attempts_before_showanswer_button
```
`attempts_before_showanswer_button` is another field that's specific to the Problem XBlock. We could use `getattr` for this one (and use `max_attempts` as the default value). This way, we will be able to refactor [the DnDv2 XBlock](https://github.com/openedx/xblock-drag-and-drop-v2) (at some point in the future) to use this service without any further changes to the `edx-platform` code.

Agrendalath commented on xmodule/services.py:
```diff
@@ -308,3 +312,147 @@ def _handle_deprecated_progress_event(self, block, event):
+
+        if self.used_all_attempts():
+            return True
+        # Using XBlock's is_past_due method, as it's suited for checking the
+        # submission deadline, unlike the is_past_due method of this service.
+        elif self._xblock.is_past_due():
+            return True
```
Where does the discrepancy between `ProblemFeedbackService.is_past_due` and `InheritableFieldsMixin.is_past_due` come from? Having two methods with the same name that behave differently is confusing.

Agrendalath commented on xmodule/services.py:
```diff
@@ -308,3 +312,147 @@ def _handle_deprecated_progress_event(self, block, event):
+            # This is after the 'never' check because course staff can see correctness
+            # unless the sequence/problem explicitly prevents it
+            return True
+        elif show_answer == ShowAnswer.ATTEMPTED:
+            return is_attempted or (has_due_date and past_due)
+        elif show_answer == ShowAnswer.ANSWERED:
+            # NOTE: this is slightly different from 'attempted' -- resetting the problems
+            # makes lcp.done False, but leaves attempts unchanged.
+            return is_correct
+        elif show_answer == ShowAnswer.CLOSED:
+            return closed
+        elif show_answer == ShowAnswer.FINISHED:
+            return closed or is_correct
+        elif show_answer == ShowAnswer.CORRECT_OR_PAST_DUE:
+            return is_correct or (has_due_date and past_due)
+        elif show_answer == ShowAnswer.PAST_DUE:
+            return past_due
+        elif show_answer == ShowAnswer.AFTER_SOME_NUMBER_OF_ATTEMPTS:
+            if max_attempts and required_attempts >= max_attempts:
+                required_attempts = max_attempts
+            return attempts >= required_attempts
+        elif show_answer == ShowAnswer.ALWAYS:
+            return True
+        elif show_answer == ShowAnswer.AFTER_ALL_ATTEMPTS:
+            return used_all_attempts
+        elif show_answer == ShowAnswer.AFTER_ALL_ATTEMPTS_OR_CORRECT:
+            return used_all_attempts or is_correct
+        elif show_answer == ShowAnswer.ATTEMPTED_NO_PAST_DUE:
+            return is_attempted
```
Suggestion: I think we could use [the `match` statement](https://peps.python.org/pep-0636/) to make this code more readable.

robrap commented:
[inform] I want to bring your attention to https://discuss.openedx.org/t/dark-launch-code-refactor-to-reduce-risk/15038. I am only adding this because of "refactor" in your commit comment. I have not actually reviewed this PR at all. Thanks.

mphilbrick211 commented:
@Agrendalath @robrap hi there! Is this pull request still needed / in progress?

Agrendalath commented:
cc: @Cup0fCoffee, ^

Cup0fCoffee commented:
@mphilbrick211 Let me confirm internally and I'll get back to you in a week.

robrap commented:
@mphilbrick211: I'm removing myself from this review. I'll leave it to others to answer your question about status. Thanks.
